### PR TITLE
Simplify UnpackInitializerData API

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: Simplify UnpackInitializerData API

**Motivation and Context**
- There are crashes reported due to unaligned memory access on arm, such as https://github.com/onnx/onnx/pull/3626, and #2813, there are many places in our mobile EP code still access the initializer data directly, which may cause issue later
- Plan is to move all the EP code accessing initializer data directly to use UnpackInitializerData
- This change is to simplify the API of UnpackInitializerData, to use std:vector instead of a pair of std::unique_ptr for data and a size_t for size, also add one extra UnpackInitializerData to work on internal initializer only (return error when it is used on tensor with external data)
  *  The only disadvantage of using vector is that the memory need to be initialized, which may have slight perf hit while creating session
- There will be on extra change to update all the EP code, one typical update is at `onnxruntime/core/providers/shared/utils/utils.cc`

### CODE REVIEW

File: onnxruntime/core/framework/tensorprotoutils.cc [hunk 1]
```diff
@@ -155,23 +155,23 @@ static Status GetExternalDataInfo(const ONNX_NAMESPACE::TensorProto& tensor_prot
 // This function does not unpack string_data of an initializer tensor
 static Status ReadExternalDataForTensor(const ONNX_NAMESPACE::TensorProto& tensor_proto,
                                         const ORTCHAR_T* tensor_proto_dir,
-                                        std::unique_ptr<unsigned char[]>& unpacked_tensor,
-                                        SafeInt<size_t>& tensor_byte_size) {
+                                        std::vector<uint8_t>& unpacked_tensor) {
```
Comment (edgchen1) [inline onnxruntime/core/framework/tensorprotoutils.cc:158]: is it possible to use std::byte?

File: onnxruntime/core/providers/shared/utils/utils.cc [hunk 2]
```diff
@@ -66,7 +66,13 @@ bool GetClipMinMax(const InitializedTensorSet& initializers, const Node& node,
...
         return false;
       }
-      min = GetTensorFloatData(*initializers.at(min_name))[0];
+      std::vector<uint8_t> unpacked_tensor;
+      auto status = onnxruntime::utils::UnpackInitializerData(*initializers.at(min_name), unpacked_tensor);
+      if (!status.IsOK()) {
+        LOGS(logger, ERROR) << "Error while unpack min tensor: " << status.ErrorMessage();
+        return false;
+      }
+      min = reinterpret_cast<float*>(unpacked_tensor.data())[0];
```
Comment (edgchen1) [inline onnxruntime/core/providers/shared/utils/utils.cc:75]: is this `vector<uint8_t>::data` guaranteed to be suitably aligned for floats?

Comment (gwang-msft) [reply onnxruntime/core/providers/shared/utils/utils.cc:75]: In theory the data in the vector will be aligned to std::max_align_t, which is usually 16, 
To make it more robust, we can change the `unpacked_tensor.resize(tensor_byte_size);` to something like
`unpacked_tensor = std::vector<std::byte>(tensor_byte_size, custom_alligned_allocator);`

Comment (edgchen1) [reply onnxruntime/core/providers/shared/utils/utils.cc:75]: thanks - sounds like it should be fine for now then

File: onnxruntime/core/framework/tensorprotoutils.cc (hunk 1 above)
Comment (gwang-msft) [reply onnxruntime/core/framework/tensorprotoutils.cc:158]: Unfortunately, seems CUDA does not have std::byte support yet, revert back to use uint8+t

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc [hunk 3]
```diff
@@ -315,14 +311,12 @@ float GetQuantizationScale(const InitializedTensorSet& initializers, const Node&
...
-  size_t tensor_byte_size;
+  std::vector<uint8_t> unpacked_tensor;
   const auto& zero_point_tensor = *initializers.at(node.InputDefs()[idx]->Name());
   ORT_RETURN_IF_ERROR(
-      onnxruntime::utils::UnpackInitializerData(zero_point_tensor, node.ModelPath(),
-                                                unpacked_tensor, tensor_byte_size));
+      onnxruntime::utils::UnpackInitializerData(zero_point_tensor, node.ModelPath(), unpacked_tensor));
   // Onnx quantization uses uint8 [int8 not yet supported], need to cast to int32_t used by NNAPI
-  zero_point = static_cast<int32_t>(unpacked_tensor.get()[0]);
+  zero_point = static_cast<int32_t>(unpacked_tensor[0]);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc:319]: >unpacked_tensor[0] [](http://example.com/codeflow?start=36&length=18)

Need to check that the data is not empty. Perhaps, this was the reason for crashes?

File: onnxruntime/core/framework/tensorprotoutils.cc [hunk 4]
```diff
@@ -155,23 +155,23 @@ static Status GetExternalDataInfo(const ONNX_NAMESPACE::TensorProto& tensor_prot
...
   ORT_RETURN_IF_ERROR(GetExternalDataInfo(
       tensor_proto,
       tensor_proto_dir,
       external_file_path,
       file_offset,
       tensor_byte_size));
 
-  unpacked_tensor.reset(new unsigned char[*&tensor_byte_size]);
+  unpacked_tensor.resize(tensor_byte_size);
   ORT_RETURN_IF_ERROR(onnxruntime::Env::Default().ReadFileIntoBuffer(
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/tensorprotoutils.cc:170]: >ORT_RETURN_IF_ERROR(onnxruntime::Env::Default().ReadFileIntoBuffer( [](http://example.com/codeflow?start=2&length=67)

Should we check for zero len?

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc (hunk 3 above)
Comment (gwang-msft) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc:319]: Yes, we should check the length of the buffer, but this is not the reason for the crashes

File: onnxruntime/core/framework/tensorprotoutils.cc (hunk 4 above)
Comment (gwang-msft) [reply onnxruntime/core/framework/tensorprotoutils.cc:170]: Seems zero len is fine here, same in posix version
https://github.com/microsoft/onnxruntime/blob/224380448d371937a084b3d162be7964a702fa9a/onnxruntime/core/platform/windows/env.cc#L255-L256

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Pull request merged**